### DISCUSSION TITLE
How to optimize Django's model's string representation database hits in browsable API

### DISCUSSION THREAD

**Discussion by HansBambel** [encode/django-rest-framework#9119] [Question & Answer]
I am currently trying to optimize my API and noticed that the browsable API is responsible for N+1 queries. This only happens when I add the parent-class in the child's `__str__` method.

Comparison using django-debug-toolbar:

```
f"{self.name} ({self.client.name})": default 531.44 ms (370 queries including 358 similar and 348 duplicates )
f"{self.name}": default 130.54 ms (24 queries including 12 similar and 2 duplicates )
```

My (simplified) models are the following:

```
class Client(models.Model):
    """Model for a Client."""

    name = models.CharField(max_length=128, unique=True, verbose_name="Company name")

    class Meta:
        """Set ordering."""

        ordering = ["name"]

    def __str__(self):
        """Return a good string representation for a Client."""
        return f"{self.name}"


class Budget(models.Model):
    """Model for representing a budget of a client."""

    name = models.CharField(max_length=128, verbose_name="Budget name")
    client = models.ForeignKey(Client, related_name="budgets", on_delete=models.PROTECT)

    class Meta:
        """Set ordering."""

        constraints = [models.UniqueConstraint(fields=["client", "name"], name="unique_client_budget_name_combination")]
        ordering = ["name"]

    def __str__(self):
        """Return a good string representation of a Budget."""
        # self.client.name triggers multiple database hits.
        return f"{self.name} ({self.client.name})"
```
This is the view:

```
class BudgetCreateAPIView(generics.ListCreateAPIView):
    """Overview of all Budgets. Uses one serializer for creating and another for displaying."""

    def get_serializer_class(self):
        """Use the display_serializer_class for displaying the model instance."""
        if self.request.method == "GET":
            return self.display_serializer_class
        return self.serializer_class

    queryset = Budget.objects.all().select_related("client")
    serializer_class = BudgetSerializer
    display_serializer_class = BudgetDisplaySerializer
```

I have also tried using select_related("client") in the view, but this did not help. It seems like the browsable API is not using the view's queryset.

How can I make the browsable API make more efficient? Is that even possible?

**Comment by Ehsan200** [comment]
Hello there!
I don't see the use of the browsable API in production; in my opinion, it's mainly for debugging purposes.

**Reply by HansBambel** [reply]
Hey :) 
Okay, but what would be a good way to check the performance of the queries? I can only see the output from the django-debug-toolbar in the browsable API.

**Reply by Ehsan200** [reply]
I don't know the debug toolbar what will show you but if you have access to the shell you can use this command and perform queries and see the actual query that Django runs:
See this [django-debug-toolbar-commands](https://django-debug-toolbar.readthedocs.io/en/latest/commands.html).

This [article](https://medium.com/@michael_england/debugging-query-performance-issues-when-using-the-django-orm-f05f83041c5f) is helpful too.

**Reply by HansBambel** [reply]
Thanks! I'll have a look.
